Changes in module 'nf-core/nanoplot'
--- modules/nf-core/nanoplot/main.nf
+++ modules/nf-core/nanoplot/main.nf
@@ -1,6 +1,11 @@
 process NANOPLOT {
     tag "$meta.id"
-    label 'process_low'
+    label 'regular'
+    cpus 8
+    memory '32 GB'
+    time '48h'
+    queue 'regular'
+    executor 'slurm'
 
     conda (params.enable_conda ? 'bioconda::nanoplot=1.40.0' : null)
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
@@ -22,12 +27,12 @@
 
     script:
     def args = task.ext.args ?: ''
-    def input_file = ("$ontfile".endsWith(".fastq.gz")) ? "--fastq ${ontfile}" :
-        ("$ontfile".endsWith(".txt")) ? "--summary ${ontfile}" : ''
+    def input_file = ("$ontfile".endsWith(".fq.gz")) ? "--fastq ${ontfile}" :
+        ("$ontfile".endsWith(".txt.gz")) ? "--summary ${ontfile}" : ''
     """
     NanoPlot \\
         $args \\
-        -t $task.cpus \\
+        -t 8 \\
         $input_file
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
